### DISCUSSION TITLE
Cef Sharp memory usage issue and deallocation problem

### DISCUSSION THREAD

**Discussion by arunkumaryadu** [cefsharp/CefSharp#4477] [General]
Hi Cef Community,

I am writing to bring to your attention an issue with Cef Sharp that I have been experiencing. Recently, I have noticed that Cef Sharp is consuming a large amount of virtual memory on my system, and it does not deallocate memory even after calling the chromiumWebBrowser dispose and Cef.Shutdown() function.

It seems that Cef Sharp is not properly deallocating memory, leading to memory leaks and potentially causing performance issues.
I have tried various solutions, such as disabling cache,disable-gpu-process and enabling multi-threaded message loop, but none of them have solved the issue.

I can see this issue on Cef_Demo_Application as well by Process.GetCurrentProcess().VirtualMemorySize64.
Always its taking almost 35GB virtual memory.

as of now i am using version 98.1.210.0, i have tried with latest version as well in latest version it is taking more than 70GB virtual memory.

I would like to request the Cef community that if this issue is already known issue then please provide a solution or workaround.
I believe that resolving this issue would not only benefit me but also many other developers who are using Cef Sharp.

Thank you for your attention to this matter.

Best regards,
Arun

**Comment by aimcosoftware** [comment]
Does your browsersubprocess.exe close, when you dispose?

**Reply by arunkumaryadu** [reply]
yes

**Reply by aimcosoftware** [reply]
`Process.GetCurrentProcess.VirtualMemorySize64` shows available cache, not used. For that you need `PrivateMemorySize64`.
They are also renowned for being inaccurate, to see a better estimate of used memory, look in Task Manager.

**Reply by arunkumaryadu** [reply]
I have stored cache on specific path, cache size is not that much. I am using cef just to render normal page so 40 GB virtual memory is to much for my use case. major concern is why its not getting deallocated after _chromiumWebBrowser.Dispose(); and Cef.Shutdown()

**Reply by aimcosoftware** [reply]
`VirtualMemorySize64` is total **available**, not total used. For that you need to use the other diagnostic values.

**Comment by amaitland** [comment]
In future please use [Discussions](https://github.com/cefsharp/CefSharp/discussions) to ask questions.

Additionally it also looks like you asked this on [stackoverflow](https://stackoverflow.com/questions/76034489/cef-sharp-memory-usage-issue-and-deallocation-problem). Please just ask your question **once** and wait for a reply. 

> It seems that Cef Sharp is not properly deallocating memory,

**Chromium** by design relies on the OS to free memory on process termination, see [Chromium Singelton](https://www.chromium.org/developers/coding-style/important-abstractions-and-data-structures/#singleton). `Cef.Shutdown` should flush memory/storage to disk and gracefully shutdown the `Chromium` sub processes. It's not possible to free all memory `Chromium` allocated.

- https://magpcss.org/ceforum/viewtopic.php?f=7&t=188
- https://magpcss.org/ceforum/viewtopic.php?f=6&t=16631

You can run `Chromium` in a separate process. There is an example at https://github.com/cefsharp/CefSharp.OutOfProcess

**Comment by BuslikDrev** [comment]
У меня кушает память запуск видео через javascript, когда я код запускаю через ajax несколько раз (то есть заменяю предыдущий тег video на новый полученный ajax способом).

Данная проблема относится в целом к браузеру.

Решил свою проблему тем, что перед заменой html тега video на новый, нужно к старому html тегу video присвоить src='';
```
	var i, l, m = document.body.querySelectorAll('video, audio');
	l = m.length;
	for (i = 0; i < l; ++i) {
		m[i].src = '';
	}
	document.body.innerHTML = ajaxData.innerHTML;
```
